docs/csharp/pattern-matching.md

```diff
@@ -2,31 +2,203 @@
+
+```csharp
+Square s;
+if (shape is Square)
+{
+    s = (shape as Square);
+    // use 's'
+} 
+```
+
+Language rules for pattern matching expressions help you avoid misusing
+the results of a match expression. In the example above, the variables `s`
+and `c` are only in scope and definitely assigned when the respective
+pattern match expressions have `true` results. If you try and use either
+variable in another location, your code generates compiler errors.
+
+Let's examine both of those rules in detail, beginning with scope. The variable
+`c` is in scope only in the `else` branch of the first `if` statement. The variable
+`s` is in scope in the method `ComputeArea`. That's because each 
+branch of an `if` statement establishes a separate scope for variables. However, the `if` statement
+itself does not. That means variables declared in the `if` statement are in the
+same scope as the `if` statement (the method in this case.) This behavior is not
+specific to pattern matching, but is the defined behavior for variable scopes 
+and `if` and `else` statements.
+
+The variables `c` and `s` are assigned when the respective `if` statements are true
+because of the definitely assigned when true mechanism.
+
+> [!TIP]
+> The samples in this topic use the recommended construct where
+> a pattern match `is` expression definitely assigns the match
+> variable in the `true` branch of the `if` statement.
+> You could reverse the logic by saying `if (!(shape is Square s))`
+> and the variable `s` would be definitely assigned only in the
+> `false` branch. While this is valid C#, it is not recommended
+> because it is more confusing to follow the logic.
+
+These rules mean that you are unlikely to accidentally access the result
+of a pattern match expression when that pattern was not met.
+
+## Using pattern matching `switch` statements
+
+As time goes on, you may need to support other shape types. As the number
+of conditions you are testing grows, you'll find that using the `is` pattern
+matching expressions can become cumbersome. In addition to requiring `if`
+statements on each type you want to check, the `is` expressions are limited
+to testing if the input matches a single type.
+
+As the number of conditions grows, you'll find that the `switch` pattern
+matching expressions becomes a better choice. To get ready to add more
+types of shapes to this library, let's change the existing method to use
+the `switch` pattern match expression:
+
+[!code-csharp[ComputeWithSwitchExpression](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#06_ComputeWithSwitchExpression "Compute with `switch` expression")]
+
+The pattern matching `switch` statement uses familiar syntax to developers
+that have used the traditional C-style `switch` statement. Each `case` is evaluated
+and the code beneath the condition that matches the input variable is
+executed. Code execution cannot "fall through" from one `case` expression
```

Comment:
I found this confusing and somewhat awkwardly worded. Perhaps "Code execution cannot "fall through" from one `case` expression to the next; the syntax of the `case` statement requires that each `case` end with a `break`, `return`, or `goto`.